### DISCUSSION TITLE
Compiling bt_music_receiver_audiokit example has #include "AudioBoard.h" error

### DISCUSSION THREAD

**Discussion by beaka** [pschatzmann/ESP32-A2DP#544] [Q&A] [answered]
Compiling bt_music_receiver_audiokit example has #include "AudioBoard.h" error.

`In file included from /home/x/Arduino/libraries/arduino-audio-tools/src/AudioLibs/AudioBoardStream.h:4,
                 from /home/x/.var/app/cc.arduino.IDE2/cache/.arduinoIDE-unsaved2024315-64-1oewxa7.98dt/bt_music_receiver_audiokit/bt_music_receiver_audiokit.ino:31:
/home/x/Arduino/libraries/arduino-audio-tools/src/AudioLibs/I2SCodecStream.h:2:10: fatal error: AudioBoard.h: No such file or directory
    2 | #include "AudioBoard.h"  // install audio-driver library
      |          ^~~~~~~~~~~~~~
compilation terminated.`

I have installed ESP32 V3.0rc with the arduino_audio_tools and arduino_audio_driver libraries in Arduino IDE v2.3.2 on Fedora 39

The old audio tools used to work but new computer and installs and now it doesn't..

**Comment by pschatzmann** [comment, marked as answer]
Double check your installation: AudioBoard.h is in the src directory of arduino_audio_driver library!
I guess you forgot to install the arduino_audio_driver library or something went wrong...

**Reply by beaka** [reply]
I did install the arduino_audio _driver but is thair any configuration for the ES8388 like the previous library?

**Reply by pschatzmann** [reply]
No, you would just pass the relevant board in the constructor of the AudioBoardStream: e.g. the old option 5 would use AudioKitEs8388V1

**Reply by beaka** [reply]
I copied AudioBoard.h from the src directory of arduino_audio_driver to arduino_audio_tools and it progressed but gave more .h errors so I copied the whole contents of arduino_audio_driver /src over and it progressed to now I am getting 
`In file included from /home/bruce/Arduino/libraries/arduino-audio-tools/src/AudioHttp/AudioHttp.h:8,
                 from /home/bruce/Arduino/libraries/arduino-audio-tools/src/AudioTools.h:94,
                 from /home/bruce/.var/app/cc.arduino.IDE2/cache/.arduinoIDE-unsaved2024315-64-1oewxa7.98dt/bt_music_receiver_audiokit/bt_music_receiver_audiokit.ino:3:
/home/bruce/Arduino/libraries/arduino-audio-tools/src/AudioHttp/URLStream.h:272:3: error: 'WiFiClient' does not name a type
  272 |   WiFiClient* clientInsecure = nullptr; // wifi client for http`

Any suggestions?

**Reply by pschatzmann** [reply]
I have tested this combination in Linux Mint (esp32 V3.0alpha, ESP32-A2DP, audio-tools, audio-driver).
So there must be something wrong with your installation...

**Reply by beaka** [reply]
Is it only necessary to copy the extracted zips of the git to the libraries folder for them to work?
I tried git clone but it keeps erroring out so I got the zips.

`Cloning into 'arduino-audio-tools'...
remote: Enumerating objects: 98768, done.
remote: Counting objects: 100% (3585/3585), done.
remote: Compressing objects: 100% (458/458), done.
error: RPC failed; curl 92 HTTP/2 stream 5 was not closed cleanly: CANCEL (err 8)
error: 4136 bytes of body are still expected
fetch-pack: unexpected disconnect while reading sideband packet
fatal: early EOF
fatal: fetch-pack: invalid index-pack output
bruce@fedora:~/Arduino/libraries$ 
`

**Reply by pschatzmann** [reply]
I guess, if you double check that the library structure is right and you have all files

**Reply by pschatzmann** [reply]
For the git issues: [this might help](https://stackoverflow.com/questions/66366582/github-unexpected-disconnect-while-reading-sideband-packet)

**Reply by beaka** [reply]
I managed to clone them with the solution in the git link, but now back to the original issue:
`In file included from /home/bruce/Arduino/libraries/arduino-audio-tools/src/AudioLibs/AudioBoardStream.h:4,
                 from /home/bruce/.var/app/cc.arduino.IDE2/cache/.arduinoIDE-unsaved2024316-60-1aqx09u.bdwp/bt_music_receiver_audiokit/bt_music_receiver_audiokit.ino:20:
/home/bruce/Arduino/libraries/arduino-audio-tools/src/AudioLibs/I2SCodecStream.h:2:10: fatal error: AudioBoard.h: No such file or directory
    2 | #include "AudioBoard.h"  // install audio-driver library
      |          ^~~~~~~~~~~~~~
compilation terminated.
exit status 1`

Will try on Ubuntu with a fresh install tomorrow.

**Reply by thierryd31** [reply]
Same problem with ESP-A1S audio kit -> board is ESP32 Wroper Module on Arduino Ide 

In file included from c:\Users\thier\Documents\Arduino\libraries\arduino-audio-tools\src/AudioTools/AudioLibs/AudioBoardStream.h:4,
                 from C:\Users\thier\Documents\Arduino\libraries\arduino-audio-tools\examples\examples-stream\streams-generator_fromarray-analog\streams-generator_fromarray-analog.ino:12:
c:\Users\thier\Documents\Arduino\libraries\arduino-audio-tools\src/AudioTools/AudioLibs/I2SCodecStream.h:2:10: fatal error: AudioBoard.h: No such file or directory
    2 | #include "AudioBoard.h"  // install audio-driver library
      |          ^~~~~~~~~~~~~~
compilation terminated.
exit status 1

Compilation error: exit status 1

**Reply by nvaken** [reply]
Also same issue here, in PlatformIO. Using the following ini.

```
[env:esp32dev]
platform = https://github.com/platformio/platform-espressif32.git
board = esp32dev
framework = arduino
lib_deps = 
    https://github.com/pschatzmann/arduino-audio-tools.git
    https://github.com/pschatzmann/arduino-audiokit.git
    https://github.com/pschatzmann/arduino-libvorbis-tremor.git
    https://github.com/pschatzmann/arduino-libopus.git
build_flags = -DCORE_DEBUG_LEVEL=5 -Wno-unused-variable -Wno-unused-but-set-variable -Wno-unused-function -Wno-format-extra-args -DCORE_DEBUG_LEVEL=5 -DAUDIOKIT_BOARD=5
monitor_speed = 115200
monitor_filters = esp32_exception_decoder
board_build.partitions = huge_app.csv
```

**Reply by pschatzmann** [reply]
Same answer as above: You are missing the arduino_audio_driver library!

**Comment by beaka** [comment]
Just did a new install of Arduino 2.3.2, esp32 v3rc, git cloned the
libraries into ~/Arduino/libraries and got the "Audiotools.h" not found. So
I copied the libraries over to ~/Documents/Arduino/libraries and it got
past that error. Is it a hardcoded path in the lib?

It now comes up with

In file included from
/home/bruce/Arduino/libraries/arduino-audio-tools/src/AudioHttp/AudioHttp.h:8,
                 from
/home/bruce/Arduino/libraries/arduino-audio-tools/src/AudioTools.h:94,
                 from
/home/bruce/.var/app/cc.arduino.IDE2/cache/.arduinoIDE-unsaved2024316-59-111tec8.drxp/bt_music_receiver_audiokit/bt_music_receiver_audiokit.ino:19:
/home/bruce/Arduino/libraries/arduino-audio-tools/src/AudioHttp/URLStream.h:272:3:
error: 'WiFiClient' does not name a type
  272 |   WiFiClient* clientInsecure = nullptr; // wifi client for http
      |   ^~~~~~~~~~
/home/bruce/Arduino/libraries/arduino-audio-tools/src/AudioHttp/URLStream.h:272:3:
note: the macro 'WiFiClient' had not yet been defined
In file included from
/home/bruce/.arduino15/packages/esp32/hardware/esp32/3.0.0-rc1/libraries/WiFi/src/WiFi.h:38,
                 from
/home/bruce/Arduino/libraries/arduino-audio-tools/src/AudioHttp/AudioServer.h:10,
                 from
/home/bruce/Arduino/libraries/arduino-audio-tools/src/AudioHttp/AudioHttp.h:10:
/home/bruce/.arduino15/packages/esp32/hardware/esp32/3.0.0-rc1/libraries/WiFi/src/WiFiClient.h:3:
note: it was later defined here
    3 | #define WiFiClient NetworkClient
      |
/home/bruce/Arduino/libraries/arduino-audio-tools/src/AudioHttp/URLStream.h:
In destructor 'virtual audio_tools::URLStream::~URLStream()':
/home/bruce/Arduino/libraries/arduino-audio-tools/src/AudioHttp/URLStream.h:57:9:
error: 'clientInsecure' was not declared in this scope; did you mean
'clientSecure'?
   57 |     if (clientInsecure != nullptr) {
      |         ^~~~~~~~~~~~~~
      |         clientSecure
/home/bruce/Arduino/libraries/arduino-audio-tools/src/AudioHttp/URLStream.h:58:14:
error: type '<type error>' argument given to 'delete', expected pointer
   58 |       delete clientInsecure;
      |              ^~~~~~~~~~~~~~
/home/bruce/Arduino/libraries/arduino-audio-tools/src/AudioHttp/URLStream.h:
In member function 'Client& audio_tools::URLStream::getClient(bool)':
/home/bruce/Arduino/libraries/arduino-audio-tools/src/AudioHttp/URLStream.h:371:9:
error: 'clientInsecure' was not declared in this scope; did you mean
'clientSecure'?
  371 |     if (clientInsecure == nullptr) {
      |         ^~~~~~~~~~~~~~
      |         clientSecure
/home/bruce/Arduino/libraries/arduino-audio-tools/src/AudioHttp/URLStream.h:372:28:
error: expected type-specifier before 'WiFiClient'
  372 |       clientInsecure = new WiFiClient();
      |                            ^~~~~~~~~~
/home/bruce/Arduino/libraries/arduino-audio-tools/src/AudioHttp/URLStream.h:375:13:
error: 'clientInsecure' was not declared in this scope; did you mean
'clientSecure'?
  375 |     return *clientInsecure;
      |             ^~~~~~~~~~~~~~
      |             clientSecure
/home/bruce/Arduino/libraries/arduino-audio-tools/src/AudioHttp/URLStream.h:
In member function 'bool audio_tools::URLStream::login()':
/home/bruce/Arduino/libraries/arduino-audio-tools/src/AudioHttp/URLStream.h:398:9:
error: 'WiFi' was not declared in this scope
  398 |         WiFi.status() != WL_CONNECTED) {
      |         ^~~~
/home/bruce/Arduino/libraries/arduino-audio-tools/src/AudioHttp/URLStream.h:398:26:
error: 'WL_CONNECTED' was not declared in this scope
  398 |         WiFi.status() != WL_CONNECTED) {
      |                          ^~~~~~~~~~~~
/home/bruce/Arduino/libraries/arduino-audio-tools/src/AudioHttp/URLStream.h:409:12:
error: 'WiFi' was not declared in this scope
  409 |     return WiFi.status() == WL_CONNECTED;
      |            ^~~~
/home/bruce/Arduino/libraries/arduino-audio-tools/src/AudioHttp/URLStream.h:409:29:
error: 'WL_CONNECTED' was not declared in this scope
  409 |     return WiFi.status() == WL_CONNECTED;
      |                             ^~~~~~~~~~~~

exit status 1

Compilation error: exit status 1

**Reply by pschatzmann** [reply]
No, there is not hardcoded path in the library. It is supposed to be copied into your libraries folder:
The destination of the Arduino Sketchbook location and the corresponding libraries folder is defined in the Arduino Preferences.

I think you still must have messed up with the versions somehow: Double check if you have some __old copies__ somewhere in the libraries directory. This error was [corrected 3 days ago](https://github.com/pschatzmann/arduino-audio-tools/commit/f7edf233b74bfea5d0b598c02faa252daa252087).

What do you get when you type 

    cd ~/Documents/Arduino/libraries/arduino-audio-tools
    git log

What board did you select when you compile ?

**Reply by pschatzmann** [reply]
I also recommend to set the flags Show verbose logs during compile and upload in the preferences, so that you can see what libraries are used in the log

**Comment by beaka** [comment]
I selected ESP32 Dev Module in all cases.

I rolled back the installation to esp32 v2.11 board and installed Arduino esp32 boards and it worked.

`
Successfully created esp32 image.
python3 /home/bruce/.arduino15/packages/esp32/hardware/esp32/2.0.11/tools/gen_esp32part.py -q /home/bruce/.var/app/cc.arduino.IDE2/cache/arduino/sketches/E2BD7FD82971CA21347A4D4D7039561B/partitions.csv /home/bruce/.var/app/cc.arduino.IDE2/cache/arduino/sketches/E2BD7FD82971CA21347A4D4D7039561B/bt_music_receiver_audiokit.ino.partitions.bin
bash -c "[ ! -d \"/home/bruce/.var/app/cc.arduino.IDE2/cache/arduino/sketches/E2BD7FD82971CA21347A4D4D7039561B\"/libraries/Insights ] || python3 \"/home/bruce/.arduino15/packages/esp32/hardware/esp32/2.0.11\"/tools/gen_insights_package.py /home/bruce/.var/app/cc.arduino.IDE2/cache/arduino/sketches/E2BD7FD82971CA21347A4D4D7039561B bt_music_receiver_audiokit.ino \"/home/bruce/.var/app/cc.arduino.IDE2/cache/.arduinoIDE-unsaved2024316-62-uf43ly.x4lmg/bt_music_receiver_audiokit\""
bash -c "[ 0 -eq 0 ] || cp -f \"/home/bruce/.arduino15/packages/esp32/tools/openocd-esp32/v0.11.0-esp32-20221026/share/openocd/scripts/\"board/esp32-wrover-kit-3.3v.cfg \"/home/bruce/.var/app/cc.arduino.IDE2/cache/.arduinoIDE-unsaved2024316-62-uf43ly.x4lmg/bt_music_receiver_audiokit\"/debug.cfg"
bash -c "[ 0 -eq 0 ] || cp -f \"/home/bruce/.arduino15/packages/esp32/hardware/esp32/2.0.11\"/tools/ide-debug/esp32.json \"/home/bruce/.var/app/cc.arduino.IDE2/cache/.arduinoIDE-unsaved2024316-62-uf43ly.x4lmg/bt_music_receiver_audiokit\"/debug_custom.json"
bash -c "[ 0 -eq 0 ] || cp -f \"/home/bruce/.arduino15/packages/esp32/hardware/esp32/2.0.11\"/tools/ide-debug/svd/esp32.svd \"/home/bruce/.var/app/cc.arduino.IDE2/cache/.arduinoIDE-unsaved2024316-62-uf43ly.x4lmg/bt_music_receiver_audiokit\"/debug.svd"

Using library audio-tools at version 0.9.6 in folder: /home/bruce/Arduino/libraries/arduino-audio-tools 
Using library WiFiClientSecure at version 2.0.0 in folder: /home/bruce/.arduino15/packages/esp32/hardware/esp32/2.0.11/libraries/WiFiClientSecure 
Using library WiFi at version 2.0.0 in folder: /home/bruce/.arduino15/packages/esp32/hardware/esp32/2.0.11/libraries/WiFi 
Using library audio-driver at version 0.0.1 in folder: /home/bruce/Arduino/libraries/arduino-audio-driver 
Using library SPI at version 2.0.0 in folder: /home/bruce/.arduino15/packages/esp32/hardware/esp32/2.0.11/libraries/SPI 
Using library Wire at version 2.0.0 in folder: /home/bruce/.arduino15/packages/esp32/hardware/esp32/2.0.11/libraries/Wire 
Using library ESP32-A2DP at version 1.8.1 in folder: /home/bruce/Arduino/libraries/ESP32-A2DP 
/home/bruce/.arduino15/packages/esp32/tools/xtensa-esp32-elf-gcc/esp-2021r2-patch5-8.4.0/bin/xtensa-esp32-elf-size -A /home/bruce/.var/app/cc.arduino.IDE2/cache/arduino/sketches/E2BD7FD82971CA21347A4D4D7039561B/bt_music_receiver_audiokit.ino.elf
Sketch uses 1368905 bytes (43%) of program storage space. Maximum is 3145728 bytes.
Global variables use 49336 bytes (15%) of dynamic memory, leaving 278344 bytes for local variables. Maximum is 327680 bytes.
`

Also had to increase memory but that did not stop compilation.
![Screenshot from 2024-04-16 20-22-02](https://github.com/pschatzmann/ESP32-A2DP/assets/778018/22dba0dc-4dd7-44ab-97ee-a2cb8120f84a)

**Reply by pschatzmann** [reply]
It is still strange that esp32 v3rc does not work for you!

**Reply by beaka** [reply]
Now that it has compiled I will see if the upgrade to V3rc works, just trying it out on the actual board. I should have said it compiles, not works yet.

**Reply by beaka** [reply]
I can upload it to the audiokit but I do not get any bluetooth device shoing up to connect to.
Sorry if being a pain, I have not coded for 15 years and am a bit rusty on this stuff.

How do I connect the device to get sound out? Using Gnome desktop Fedora 39, stock standard. Blurtooth is working as I use BT mouse an kbd.

**Comment by beaka** [comment]
Got it to list in the BT devices and connected.

Playing video does not get sound through headphones on audiokit so am now
checking the API to set the headphone output.

The A2DP sink is selected in the Sound settings.

**Comment by pschatzmann** [comment]
The earphones should work automatically: the default setting is using output_device = DAC_OUTPUT_ALL.
Maybe it is a question of the volume ?
Try with some low impedance (16-32 ohm) earphones...

**Comment by beaka** [comment]
Just went through the API and found output_device = DAC_OUTPUT_ALL.

All volumes are up and sound test from Gnome settings does not produce any
sound.

Will keep trying.

I have had sound out of the board before, working on some FIR's and
Hilberts for an SDR  but will try another board in case my inexperience did
something.

**Comment by beaka** [comment]
Yay got BT sound through the Audiokit. Had to change the Configuration in
Sound Settings to (A2DP Sink, Codec SBC) from (A2DP Sink, Codec SBC-XQ)
which was selected automatically.

Not HiFi sound though, lacks the "fullness" of the Sennheiser headphones
plugged directly into the tablet but clear and not noisy or distorted.

Will try compiling with the V3rc tomorrow, after the frustration levels
subside.
